src/cata_utility.cpp

```diff
@@ -315,3 +317,79 @@ float multi_lerp( const std::vector<std::pair<float, float>> &points, float x )
     // Linear interpolation of values at relevant thresholds
     return ( t * points[i].second ) + ( ( 1 - t ) * points[i - 1].second );
 }
+
+ofstream_wrapper::ofstream_wrapper( const std::string &path )
+{
+    file_stream.open( path.c_str(), std::ios::binary );
+    if( !file_stream.is_open() ) {
+        throw std::runtime_error( "opening file failed" );
+    }
+}
+
+ofstream_wrapper::~ofstream_wrapper() = default;
+
+void ofstream_wrapper::close()
+{
+    file_stream.close();
+    if( file_stream.fail() ) {
+        throw std::runtime_error( "writing to file failed" );
+    }
+}
+
+bool write_to_file( const std::string &path, const std::function<void( std::ostream & )> &writer,
+                    const char *const fail_message )
+{
+    try {
+        ofstream_wrapper fout( path );
+        writer( fout.stream() );
+        fout.close();
+        return true;
+
+    } catch( const std::exception &err ) {
+        if( fail_message ) {
+            popup( _( "Failed to write %1$s to \"%2$s\": %3$s" ), fail_message, path.c_str(), err.what() );
```

Comment:
The user may encounter this message (file system is full, or insufficient permissions to write files), they indicate a problem on the their system, not necessarily in the program. They need to be useful for the user.

Debug messages are not supposed to appear at all and indicate a problem in the program, they need to be useful for the programmer (and the user is supposed to forward them verbatim).

I made the exception messages translatable as well.